From: Michael Olbrich <m.olbrich@pengutronix.de>
Date: Wed, 24 Apr 2013 15:56:52 +0200
Subject: [PATCH] v4l2: filter: reorder device configuration and activation

Call streamon on the v4l2 output side before negotiating a format on the
capture side. This allows the kernel driver to limit the capture side
formats to the ones possible given the already fixed format on the output
side.

Signed-off-by: Michael Olbrich <m.olbrich@pengutronix.de>
---
 sys/v4l2/gstv4l2filter.c | 37 +++++++++++++++++++++++--------------
 1 file changed, 23 insertions(+), 14 deletions(-)

diff --git a/sys/v4l2/gstv4l2filter.c b/sys/v4l2/gstv4l2filter.c
index 523f2a0..e19c19b 100644
--- a/sys/v4l2/gstv4l2filter.c
+++ b/sys/v4l2/gstv4l2filter.c
@@ -446,9 +446,6 @@ gst_v4l2filter_sink_setcaps (GstPad * pad, GstCaps * caps)
 
   GST_INFO_OBJECT (v4l2filter, "setting sink caps: %" GST_PTR_FORMAT, caps);
 
-  if (!gst_v4l2filter_src_update_caps (v4l2filter, caps))
-    return FALSE;
-
   return TRUE;
 
   /* ERRORS */
@@ -740,16 +737,6 @@ gst_v4l2filter_sink_chain (GstPad * pad, GstObject * parent, GstBuffer * buffer)
   if (G_UNLIKELY (obj->pool == NULL))
     goto not_negotiated;
 
-  GST_V4L2FILTER_PREROLL_LOCK (v4l2filter);
-  if (GST_PAD_MODE (v4l2filter->srcpad) != GST_PAD_MODE_NONE)
-    gst_v4l2filter_start_task (v4l2filter);
-  GST_V4L2FILTER_PREROLL_UNLOCK (v4l2filter);
-
-  if (!GST_PAD_TASK (v4l2filter->srcpad)) {
-    gst_buffer_unref (buffer);
-    return GST_FLOW_FLUSHING;
-  }
-
   time = g_slice_new0 (GstV4l2FilterTime);
   time->pts = GST_BUFFER_PTS (buffer);
   time->dts = GST_BUFFER_DTS (buffer);
@@ -773,7 +760,29 @@ gst_v4l2filter_sink_chain (GstPad * pad, GstObject * parent, GstBuffer * buffer)
       buffer);
   gst_buffer_unref (buffer);
 
-  return ret;
+  if (G_UNLIKELY (ret != GST_FLOW_OK))
+    return ret;
+
+  if (!GST_PAD_TASK (v4l2filter->srcpad)) {
+    GstStructure *s;
+    GstCaps *caps;
+    s = gst_buffer_pool_get_config (GST_BUFFER_POOL_CAST (obj->pool));
+    gst_buffer_pool_config_get_params (s, &caps, NULL, NULL, NULL);
+    if (!gst_v4l2filter_src_update_caps (v4l2filter, caps))
+      return GST_FLOW_FLUSHING;
+  }
+
+  GST_V4L2FILTER_PREROLL_LOCK (v4l2filter);
+  if (GST_PAD_MODE (v4l2filter->srcpad) != GST_PAD_MODE_NONE)
+    gst_v4l2filter_start_task (v4l2filter);
+  GST_V4L2FILTER_PREROLL_UNLOCK (v4l2filter);
+
+  if (!GST_PAD_TASK (v4l2filter->srcpad)) {
+    gst_buffer_unref (buffer);
+    return GST_FLOW_FLUSHING;
+  }
+
+  return GST_FLOW_OK;
 
   /* ERRORS */
 not_negotiated:
